bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/TapoCredentials.java

```diff
@@ -0,0 +1,186 @@
+    }
+
+    /**
+     * set credentials.
+     *
+     * @param username username (eMail-adress) of Tapo Cloud
+     * @param passowrd Password of Tapo Cloud
+     */
+    public void setCredectials(String eMail, String password) {
+        try {
+            encryptCredentials(eMail, password);
+            createKeyPair();
+        } catch (Exception e) {
+            logger.trace("error init credential class '{}'", e.toString());
+        }
+    }
+
+    /**
+     * encrypt credentials.
+     *
+     * @param username username (eMail-adress) of Tapo Cloud
+     * @param passowrd Password of Tapo Cloud
+     */
+    private void encryptCredentials(String username, String password) throws Exception {
+        logger.trace("encrypt credentials for '{}'", username);
+
+        /* Password Encoding */
+        byte[] byteWord = password.getBytes();
+        this.encodedPassword = mimeEncoder.encodeToString(byteWord);
+
+        /* User Encoding */
+        String encodedUser = this.shaDigestUsername(username);
+        byteWord = encodedUser.getBytes("UTF-8");
+        this.encodedEmail = mimeEncoder.encodeToString(byteWord);
+    }
+
+    /**
+     * Create Key-Pairs
+     *
+     */
+    public void createKeyPair() throws NoSuchAlgorithmException {
+        KeyPairGenerator instance = KeyPairGenerator.getInstance("RSA");
+        instance.initialize(1024, new SecureRandom());
+        KeyPair generateKeyPair = instance.generateKeyPair();
+
+        this.publicKey = new String(
+                java.util.Base64.getMimeEncoder().encode(((RSAPublicKey) generateKeyPair.getPublic()).getEncoded()));
+        this.privateKey = new String(
+                java.util.Base64.getMimeEncoder().encode(((RSAPrivateKey) generateKeyPair.getPrivate()).getEncoded()));
+    }
+
+    /**
+     * shaDigest USERNAME
+     *
+     */
+    private String shaDigestUsername(String str) throws NoSuchAlgorithmException {
+        byte[] bArr = str.getBytes();
+        byte[] digest = MessageDigest.getInstance("SHA1").digest(bArr);
+
+        StringBuilder sb = new StringBuilder();
+        for (byte b : digest) {
+            String hexString = Integer.toHexString(b & 255);
+            if (hexString.length() == 1) {
+                sb.append("0");
+                sb.append(hexString);
+            } else {
+                sb.append(hexString);
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * RETURN ENCODED PASSWORD
+     *
+     */
+    public String getEncodedPassword() {
+        return encodedPassword;
+    }
+
+    /**
+     * RETURN ENCODED E-MAIL
+     *
+     */
+    public String getEncodedEmail() {
+        return encodedEmail;
+    }
+
+    /**
+     * RETURN PRIVATE-KEY
+     * 
+     * @return String -----BEGIN PRIVATE KEY-----\n%s\n-----END PRIVATE KEY-----
+     */
+    public String getPrivateKey() {
+        return String.format("-----BEGIN PRIVATE KEY-----\n%s\n-----END PRIVATE KEY-----\n", privateKey);
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.